File: src/streams.py
HR_ZONES = [(0, 130), (130, 148), (148, 162), (162, 174), (174, 9999)]


def compute_streams_derived(streams: dict, activity: dict) -> dict:
    hr_data = streams.get("heartrate", {}).get("data", [])
    alt_data = streams.get("altitude", {}).get("data", [])
    vel_data = streams.get("velocity_smooth", {}).get("data", [])
    grade_data = streams.get("grade_smooth", {}).get("data", [])
    cadence_data = streams.get("cadence", {}).get("data", [])

    n_hr = len(hr_data)

    # Elevation loss (sum of negative altitude deltas)
    elevation_loss = 0.0
    for i in range(1, len(alt_data)):
        delta = alt_data[i] - alt_data[i - 1]
        if delta < 0:
            elevation_loss += abs(delta)

    # Time in HR zones
    zone_counts = [0] * 5
    for hr in hr_data:
        for i, (lo, hi) in enumerate(HR_ZONES):
            if lo <= hr < hi:
                zone_counts[i] += 1
                break
    pct_zones = [round(c / n_hr * 100, 1) if n_hr > 0 else 0.0 for c in zone_counts]

    # Aerobic decoupling (pace:HR efficiency ratio, first half vs second half)
    decoupling_pct = None
    if hr_data and vel_data and n_hr >= 20 and len(vel_data) == n_hr:
        mid = n_hr // 2
        hr_first = sum(hr_data[:mid]) / mid
        hr_second = sum(hr_data[mid:]) / (n_hr - mid)
        v_first = sum(vel_data[:mid]) / mid
        v_second = sum(vel_data[mid:]) / (n_hr - mid)
        ratio_first = v_first / hr_first if hr_first > 0 else 0
        ratio_second = v_second / hr_second if hr_second > 0 else 0
        if ratio_first > 0:
            decoupling_pct = round((ratio_second - ratio_first) / ratio_first * 100, 2)

    # Grade-adjusted pace (average, converted to min/km)
    gap = None
    if vel_data and grade_data and len(vel_data) == len(grade_data):
        adjusted_velocities = []
        for v, g in zip(vel_data, grade_data):
            if v > 0:
                # Each 1% grade ≈ +3.3% energy cost (Jack Daniels approximation)
                grade_factor = 1.0 + 0.033 * g
                adjusted_v = v * grade_factor
                adjusted_velocities.append(adjusted_v)
        if adjusted_velocities:
            avg_v = sum(adjusted_velocities) / len(adjusted_velocities)
            gap = round(1000 / 60 / avg_v, 2) if avg_v > 0 else None

    # Cadence (double the Strava half-cadence to get full SPM)
    cadence_avg = None
    if cadence_data:
        cadence_avg = round(sum(cadence_data) / len(cadence_data) * 2, 1)

    return {
        "activity_id": activity["id"],
        "elevation_loss_m": round(elevation_loss, 1),
        "decoupling_pct": decoupling_pct,
        "pct_time_z1": pct_zones[0],
        "pct_time_z2": pct_zones[1],
        "pct_time_z3": pct_zones[2],
        "pct_time_z4": pct_zones[3],
        "pct_time_z5": pct_zones[4],
        "grade_adjusted_pace": gap,
        "cadence_avg": cadence_avg,
    }

File: web/src/lib/strava/streams.ts
// Ported from src/streams.py.
import type { StreamsDerivedInput } from "../db/mutations";

export interface StravaStreamsResponse {
  heartrate?: { data?: number[] };
  altitude?: { data?: number[] };
  velocity_smooth?: { data?: number[] };
  grade_smooth?: { data?: number[] };
  cadence?: { data?: number[] };
}

export function computeStreamsDerived(
  streams: StravaStreamsResponse,
  activityId: number,
  hrZones: Array<[number, number]>,
): StreamsDerivedInput {
  const hrData = streams.heartrate?.data ?? [];
  const altData = streams.altitude?.data ?? [];
  const velData = streams.velocity_smooth?.data ?? [];
  const gradeData = streams.grade_smooth?.data ?? [];
  const cadenceData = streams.cadence?.data ?? [];

  const nHr = hrData.length;

  let elevationLoss = 0.0;
  for (let i = 1; i < altData.length; i++) {
    const delta = altData[i] - altData[i - 1];
    if (delta < 0) elevationLoss += Math.abs(delta);
  }

  const zoneCounts = new Array(hrZones.length).fill(0);
  for (const hr of hrData) {
    for (let i = 0; i < hrZones.length; i++) {
      const [lo, hi] = hrZones[i];
      if (hr >= lo && hr < hi) {
        zoneCounts[i] += 1;
        break;
      }
    }
  }
  const pctZones = zoneCounts.map((c) => (nHr > 0 ? Math.round((c / nHr) * 100 * 10) / 10 : 0.0));

  let decouplingPct: number | null = null;
  if (hrData.length > 0 && velData.length > 0 && nHr >= 20 && velData.length === nHr) {
    const mid = Math.floor(nHr / 2);
    const hrFirst = hrData.slice(0, mid).reduce((a, b) => a + b, 0) / mid;
    const hrSecond = hrData.slice(mid).reduce((a, b) => a + b, 0) / (nHr - mid);
    const vFirst = velData.slice(0, mid).reduce((a, b) => a + b, 0) / mid;
    const vSecond = velData.slice(mid).reduce((a, b) => a + b, 0) / (nHr - mid);
    const ratioFirst = hrFirst > 0 ? vFirst / hrFirst : 0;
    const ratioSecond = hrSecond > 0 ? vSecond / hrSecond : 0;
    if (ratioFirst > 0) {
      decouplingPct = Math.round(((ratioSecond - ratioFirst) / ratioFirst) * 100 * 100) / 100;
    }
  }

  let gap: number | null = null;
  if (velData.length > 0 && gradeData.length > 0 && velData.length === gradeData.length) {
    const adjustedVelocities: number[] = [];
    for (let i = 0; i < velData.length; i++) {
      const v = velData[i];
      const g = gradeData[i];
      if (v > 0) {
        // Each 1% grade ≈ +3.3% energy cost (Jack Daniels approximation)
        const gradeFactor = 1.0 + 0.033 * g;
        adjustedVelocities.push(v * gradeFactor);
      }
    }
    if (adjustedVelocities.length > 0) {
      const avgV = adjustedVelocities.reduce((a, b) => a + b, 0) / adjustedVelocities.length;
      gap = avgV > 0 ? Math.round((1000 / 60 / avgV) * 100) / 100 : null;
    }
  }

  let cadenceAvg: number | null = null;
  if (cadenceData.length > 0) {
    cadenceAvg = Math.round((cadenceData.reduce((a, b) => a + b, 0) / cadenceData.length) * 2 * 10) / 10;
  }

  return {
    activity_id: activityId,
    elevation_loss_m: Math.round(elevationLoss * 10) / 10,
    decoupling_pct: decouplingPct,
    pct_time_z1: pctZones[0] ?? 0.0,
    pct_time_z2: pctZones[1] ?? 0.0,
    pct_time_z3: pctZones[2] ?? 0.0,
    pct_time_z4: pctZones[3] ?? 0.0,
    pct_time_z5: pctZones[4] ?? 0.0,
    grade_adjusted_pace: gap,
    cadence_avg: cadenceAvg,
  };
}
